modules/BookingService.Orleans/src/EasyAbp.BookingService.Orleans.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyGrain.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using EasyAbp.BookingService.AssetOccupancies;
+using EasyAbp.BookingService.AssetSchedules;
+using Orleans;
+using Orleans.Providers;
+
+namespace EasyAbp.BookingService.AssetOccupancyProviders;
+
+[StorageProvider(ProviderName = StorageProviderName)]
+public class AssetOccupancyGrain : Grain<AssetOccupancyStateModel>, IAssetOccupancyGrain
+{
+    public const string StorageProviderName = "BookingServiceAssetOccupancyStorage";
+
+    private Guid _assetId;
+    private DateTime _date;
+
+    public override Task OnActivateAsync()
+    {
+        _assetId = this.GetPrimaryKey(out var compoundKey);
+        _date = compoundKey.GetDateFromCompoundKey();
+        return base.OnActivateAsync();
+    }
+
+    public virtual Task<List<ProviderAssetOccupancyModel>> GetAssetOccupanciesAsync()
+    {
+        return Task.FromResult(State.AssetOccupancies
+            .Select(x => new ProviderAssetOccupancyModel(_assetId, x.Volume, _date, x.StartingTime, x.Duration))
+            .ToList());
+    }
+
+    public virtual async Task<ProviderAssetOccupancyModel> OccupyAsync(ProviderOccupyingInfoModel model)
+    {
+        var assetOccupancyModel = State.AssetOccupancies.FirstOrDefault(x =>
+            x.StartingTime == model.StartingTime && x.Duration == model.Duration);
+
+        if (assetOccupancyModel is null)
+        {
+            assetOccupancyModel = new AssetOccupancyCountModel(model.StartingTime, model.Duration, model.Volume);
+            State.AssetOccupancies.Add(assetOccupancyModel);
+        }
+        else
+        {
+            if (assetOccupancyModel.Volume + model.Volume > model.Asset.Volume)
+            {
+                throw new InsufficientAssetVolumeException();
+            }
+
+            if (!assetOccupancyModel.ChangeVolume(model.Volume))
```

Comment:
I think the `ProviderOccupyingInfoModel` should check `Volume > 0`.

Reply:
Maybe negative volume can be useful when a admin is adjusting occupancy manually?

Reply:
Yes, and the `ProviderOccupyingInfoModel` is used for occupancy, which only reduces volume.